pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java

```diff
@@ -500,15 +500,24 @@ public Response downloadValidDocIds(
             String.format("Table %s segment %s is not a immutable segment", tableNameWithType, segmentName),
             Response.Status.BAD_REQUEST);
       }
-      MutableRoaringBitmap validDocIds =
-          indexSegment.getValidDocIds() != null ? indexSegment.getValidDocIds().getMutableRoaringBitmap() : null;
-      if (validDocIds == null) {
+
+      // Adopt the same logic as the query execution to get the valid doc ids. 'FilterPlanNode.run()'
+      // If the queryableDocId is available (upsert delete is enabled), we read the valid doc ids from it.
+      // Otherwise, we read the valid doc ids.
+      MutableRoaringBitmap validDocIdSnapshot = null;
```

Comment:
```suggestion
      final MutableRoaringBitmap validDocIdSnapshot;
```
I believe this must either equal `indexSegment.getQueryableDocIds().getMutableRoaringBitmap()` or `indexSegment.getValidDocIds().getMutableRoaringBitmap()` and if it not assigned one of those then an exception is thrown: so this can be changed to be immutable.

Reply:
changed to `final`